File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        arr = ["" for i in range(26)]

        for letter in letters:
            arr[ord(letter) - ord("a")] = letter
        
        for i in range(26):

            if arr[i] != "" and ord(target) < ord(arr[i]):
                return arr[i]
        
        return letters[0]

#Time complexity: O(n)
# Space complexity: O(1)

File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.java
class Solution {
    public char nextGreatestLetter(char[] letters, char target) {
        int n = letters.length;
        int low = 0;
        int high = n - 1;
        char result = letters[0];

        while(low <= high){
            int mid = low + (high - low) / 2;

            if(target < letters[mid]){
                result = letters[mid];
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }

        return result;
    }
}